### PR TITLE
cli: add `--author` argument for `commit` and `describe`

### PR DESCRIPTION
**Pull request by mati865**
# Checklist

If applicable:
- [x] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (cli/src/config-schema.json)
- [x] I have added tests to cover my changes

Fixes https://github.com/martinvonz/jj/issues/4170

### CODE REVIEW

File: cli/src/commands/commit.rs
Comment (mati865) [inline cli/src/commands/commit.rs:57]: In the issues author said:
> If more than a single --author option is specified, the operation should fail, instructing the user to specify, at most, one.

I'm not sure if this is doable with Clap.

Comment (PhilipMetzger) [reply cli/src/commands/commit.rs:57]: Since the author field is not a `Vec`, you cannot pass it multiple times so it is fine as is.

**Review by PhilipMetzger — APPROVED** [review]
LGTM, it's only missing a UI test in `cli/tests/test_describe.rs` for the feature.

File: cli/src/commands/commit.rs
Comment (mati865) [reply cli/src/commands/commit.rs:57]: I thought Clap would override an argument with the last instance by default, but maybe I have confused it with another crate.

**Comment by mati865** [comment]
Sorry, I've looked for these tests but couldn't find them previously. I'll update them tomorrow.

File: cli/src/commands/describe.rs
Comment (mati865) [inline cli/src/commands/describe.rs:248]: https://github.com/martinvonz/jj/blob/56dbbb8fc629a80173b1253e6baec225bc83b8f6/lib/src/commit_builder.rs#L208 modifies the timestamp, so I'm reading it before `rewritter.rebase` is called.

File: cli/src/commands/describe.rs
Comment (mati865) [inline cli/src/commands/describe.rs:128]: `else` branch is lacking tests for both `--reset-author` and `--author`

File: cli/src/description_util.rs
Comment (mati865) [inline cli/src/description_util.rs:255]: I hadn't thought about that earlier, but it could be moved to Clap `https://docs.rs/clap/latest/clap/struct.Arg.html#method.value_parser`

**Review by mati865 — COMMENTED** [review]
The tests revealed a few issues that should be addressed now. I'll move value parsing to Clap and rebase the conflict later.

File: cli/src/commands/describe.rs
Comment (yuja) [reply cli/src/commands/describe.rs:248]: I think the updated timestamp should be preserved in that case. For consistency, `jj describe` should update author timestamp if the commit was empty.

Maybe adjust the doc string instead?

Comment (mati865) [reply cli/src/commands/describe.rs:246]: That would be inconsistent with Git and also lead to strange-looking history that has nonchronological timestamps.

File: cli/src/description_util.rs
Comment (mati865) [reply cli/src/description_util.rs]: Done

File: cli/src/commands/describe.rs
Comment (PhilipMetzger) [reply cli/src/commands/describe.rs:246]: `jj` always had that, see #3906 and it's associated bug (#2000). There's no reason that we have to be consistent with Git here.

Comment (mati865) [reply cli/src/commands/describe.rs]: I fully agree with regard to resetting the date of `jj new` commit, not so much about resetting timestamp of other commits but updated PR to that version.
Since this option is now the same as `JJ_USER='Foo Bar' JJ_EMAIL=foo@bar.com jj describe --reset-author` I'm not sure if  there is much benefit from it, shorter and more intuitive syntax seems like the only selling point.

Comment (PhilipMetzger) [reply cli/src/commands/describe.rs]: I think it was intended as that, so to my understanding this is also fine.

Comment (yuja) [reply cli/src/commands/describe.rs]: > Since this option is now the same as `JJ_USER='Foo Bar' JJ_EMAIL=foo@bar.com jj describe --reset-author`

`--reset-author` always updates author timestamp, but `--author` doesn't, right? I don't think Git consistency is important here, but it's somewhat consistent because in Git, we don't (usually) have a commit without description and content.

Comment (mati865) [reply cli/src/commands/describe.rs]: Sorry, my bad.

I had based this entirely on my misunderstanding of the unit test. Commits there have timestamps rewritten because they have empty descriptions, right?

Running on the real repository shows the author timestamp intact and committer timestamp updated as intended:
```
/tmp/author-test
❯ ~/Projects/jj/target/debug/jj log -T "description ++ ' ' ++ author.timestamp() ++ ' ' ++ committer.timestamp()"
@   2024-09-17 01:41:42.000 +02:00 2024-09-17 01:41:42.000 +02:00
○  bar
│   2024-09-17 01:41:42.000 +02:00 2024-09-17 01:41:42.000 +02:00
○  foo
│   2024-09-17 01:41:32.000 +02:00 2024-09-17 01:41:33.000 +02:00
◆   1970-01-01 00:00:00.000 +00:00 1970-01-01 00:00:00.000 +00:00

/tmp/author-test
❯ jj log
@  vmrplqvo mati865@gmail.com 2024-09-17 01:41:42 4da73e4d
│  (empty) (no description set)
○  ztxlzwmq mati865@gmail.com 2024-09-17 01:41:42 2919a8bd
│  bar
○  xonnomqr mati865@gmail.com 2024-09-17 01:41:33 6359b7c7
│  foo
◆  zzzzzzzz root() 00000000

/tmp/author-test
❯ ~/Projects/jj/target/debug/jj desc @-- --no-edit --author "foo bar <foo@example.com>"
Rebased 2 descendant commits
Working copy now at: vmrplqvo 7bacbeed (empty) (no description set)
Parent commit      : ztxlzwmq 9c1a7191 bar

/tmp/author-test
❯ ~/Projects/jj/target/debug/jj log -T "description ++ ' ' ++ author.timestamp() ++ ' ' ++ committer.timestamp()"
@   2024-09-17 01:47:25.000 +02:00 2024-09-17 01:47:25.000 +02:00
○  bar
│   2024-09-17 01:41:42.000 +02:00 2024-09-17 01:47:25.000 +02:00
○  foo
│   2024-09-17 01:41:32.000 +02:00 2024-09-17 01:47:25.000 +02:00
◆   1970-01-01 00:00:00.000 +00:00 1970-01-01 00:00:00.000 +00:00
```

I have updated the docstrings to indicate timestamp preservation if the description is nonempty.

File: cli/src/cli_util.rs
Comment (yuja) [inline cli/src/cli_util.rs:3302]: nit: we don't have to use `clap::Error`. `&'static str` or `String` also works.

`text_util` or `description_util` might be better place to host this function. `cli_util` is bloated. We have `split_email()` in `text_util.rs` btw.

File: cli/src/cli_util.rs
Comment (yuja) [inline cli/src/cli_util.rs:3303]: needs `$` anchor to not discard trailing text?

I think using `.rsplit_once()`, `.strip_suffix()`, `.trim_ascii()`, etc. is also good.

File: cli/src/commands/commit.rs
Comment (yuja) [inline cli/src/commands/commit.rs:57]: nit: "if the description is filled" isn't accurate. Maybe we can say "discardable" or "empty commit with no description" (copied from 1eebbe57c080568eec96e548c02ab8e43d8b1d8a)

(also applies to `cmd_describe()`)

File: cli/src/commands/commit.rs
Comment (yuja) [inline cli/src/commands/commit.rs:124]: you mean `author().timestamp`?

File: cli/src/commands/commit.rs
Comment (yuja) [inline cli/src/commands/commit.rs:121]: nit: we usually import `jj_lib::backend::Signature`

File: cli/src/commands/describe.rs
Comment (yuja) [inline cli/src/commands/describe.rs:220]: nit: strictly speaking, we should test `new_author == commit.author()` because new author might be unchanged (unlike `--reset-author`.) Perhaps, we can refactor `commit_descriptions` to a list of `(old_commit, new_author, new_description)`.

This isn't a blocker. Can be addressed later.

File: cli/tests/test_describe_command.rs
Comment (yuja) [inline cli/tests/test_describe_command.rs:535]: Did you intentionally include quotes in user name?

**Review by yuja — APPROVED** [review]
LGTM, thanks!

File: cli/src/cli_util.rs
Comment (mati865) [reply cli/src/cli_util.rs]: Right, I've got so used to `anyhow`/`thiserror` in recent years I tried to create `&'static str` error but with one `Err()` too many over the weekend.
Changed to `&'static str`.

File: cli/src/cli_util.rs
Comment (mati865) [reply cli/src/cli_util.rs]: `$` makes sense, if there is something after `>` then it's improper syntax.
Added.

File: cli/src/commands/commit.rs
Comment (mati865) [reply cli/src/commands/commit.rs]: I think you meant negation of these statements? Author timestamps of discardable commits have no value.
Changed to `non-discardable commits`.

File: cli/src/commands/commit.rs
Comment (mati865) [reply cli/src/commands/commit.rs:125]: This is intentional. using author's timestamp gives weird results:
```
-- STDOUT:              jj-cli::runner test_commit_command::test_commit_with_description_template ---

running 1 test
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ Snapshot Summary ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
Snapshot: commit_with_description_template-2
Source: cli/tests/test_commit_command.rs:244
────────────────────────────────────────────────────────────────────────────────
Expression: std::fs::read_to_string(test_env.env_root().join("editor")).unwrap()
────────────────────────────────────────────────────────────────────────────────
-old snapshot
+new results
────────────┬───────────────────────────────────────────────────────────────────
    0       │-JJ: Author: Another User <another.user@example.com> (2001-02-03 08:05:09)
          0 │+JJ: Author: Another User <another.user@example.com> (2001-02-03 08:05:08)
    1     1 │ JJ: Committer: Test User <test.user@example.com> (2001-02-03 08:05:09)
    2     2 │
    3     3 │ JJ: file2 | 1 +
    4     4 │ JJ: 1 file changed, 1 insertion(+), 0 deletions(-)
────────────┴───────────────────────────────────────────────────────────────────
To update snapshots run `cargo insta review`
Stopped on the first failure. Run `cargo insta test` to run all snapshots.
test test_commit_command::test_commit_with_description_template ... FAILED

failures:

failures:
    test_commit_command::test_commit_with_description_template

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 598 filtered out; finished in 0.89s
```
This would be inconsistent with other cases in that test.

File: cli/src/commands/describe.rs
Comment (mati865) [reply cli/src/commands/describe.rs]: This is more complicated, we need to test whether the author is none or equal, so it's either:
```rust
                && !args.author.as_ref().is_some_and(|(name, email)| {
                    name != &commit.author().name || email != &commit.author().email
                })
```
or
```rust
                && args.author.is_none()
                && args.author.as_ref().is_some_and(|(name, email)| {
                    name == &commit.author().name && email == &commit.author().email
                })
```

File: cli/tests/test_describe_command.rs
Comment (mati865) [reply cli/tests/test_describe_command.rs]: 🤦🏻‍♂️ I was fixing a script and forgot to change my mindset that I'm using sane language again.

File: cli/src/commands/commit.rs
Comment (yuja) [reply cli/src/commands/commit.rs:125]: Well, if you don't pass `--author` here, you'll get the author timestamp of "2001-02-03 08:05:08", right? (Beware that the "file2" exists when the first commit is created (or split by `jj commit file1`.) Since `--author` shouldn't affect how author timestamp is updated, the tested behavior is wrong.

File: cli/src/text_util.rs
Comment (yuja) [inline cli/src/text_util.rs:264]: nit: let's remove `_arg` as it's no longer return `clap::Error`.

File: cli/src/commands/describe.rs
Comment (yuja) [reply cli/src/commands/describe.rs]: Yep, we might want to refactor it later.

You could use `.map_or(true, ..)`, btw, but it might be slightly simpler if we invert the condition.
```rust
        .filter(|(commit, new_description)| {
            new_description != commit.description()
                || args.reset_author
                || args.author.as_ref().is_some_and(|(name, email)| {
                    name != &commit.author().name || email != &commit.author().email
                })
        })
        .map(|(commit, new_description)| (commit.id(), new_description))
```

File: cli/src/commands/commit.rs
Comment (yuja) [reply cli/src/commands/commit.rs]: Looks good, thanks.

File: cli/src/commands/commit.rs
Comment (mati865) [reply cli/src/commands/commit.rs]: Ah, that explains it.
Changed to author timestamp.

**Comment by mati865** [comment]
Thank you for bearing with me through the whole process!
I'll consider how to refactor problematic code in `describe.rs` but probably not until next weekend.

File: cli/src/commands/describe.rs
Comment (yuja) [reply cli/src/commands/describe.rs]: > I'll consider how to refactor problematic code in describe.rs but probably not until next weekend.

Oops, I didn't mean this was a blocker. I think the current state is good, and this PR can be merged if you're okay.

Comment (mati865) [reply cli/src/commands/describe.rs]: Oh, sorry. I have badly worded that sentence.
I mean I'll investigate it but likely not before next weekend, and I've enabled auto-merge. The last time I used GH auto merge, the conversations weren't a blocker, so I'll resolve them.

**Comment by mati865** [comment]
I couldn't find the way to refactor it in a way that would improve it, felt like I'm only making things worse.

**Comment by omentic** [comment]
This flag appears to be undocumented.

**Comment by mati865** [comment]
@omentic could you be more verbose about the issue? I think (cannot test ATM) it should show up in the `--help` for the mentioned args.
I'm happy to extend the jj  docs if that's what you mean.

**Comment by bnjmnt4n** [comment]
The feature has been deprecated in favor of `jj metaedit`: https://github.com/jj-vcs/jj/commit/51224ed675dfc6430b35f6cdf0c2fb085b3107be.

**Comment by omentic** [comment]
@mati865 It doesn't show up in `--help`, yeah. Thanks for the info @bnjmnt4n -- I was playing around with flags and noticed this `--author` existed, but I guess I didn't actually *run* it to get the deprecation warning (which appears).

**Pull request merged**
